### PR TITLE
Match work_item_id in visualiser search

### PR DESCRIPTION
**Pull request by phelma**
## Summary

Extends visualiser search to match on `work_item_id`, alongside related work-item tooling and test hardening accumulated on this branch.

## Changes

- **Visualiser search** — match `work_item_id` in search; update indexer and add `api_search` integration tests (~500 lines).
- **Work items** — link 0124–0156 to Linear and refresh the sync baseline.
- **Prevention tests** — agent-invocation prevention tests for 0120 (research, plan, review); harden the no-input migration stall regression test; cross-check tolerant unknown backfill against the corpus validator.

## Test plan

- [ ] `mise run check`
- [ ] `cd skills/visualisation/visualise/server && cargo test`

### CODE REVIEW

File: skills/visualisation/visualise/server/src/api/search.rs [hunk 1]
```diff
@@ -18,6 +18,12 @@ use crate::server::AppState;
 /// whitespace cannot be used to pad past the cap.
 const MAX_Q_LEN: usize = 128;
 
+/// Hard cap on returned rows. Bounds the response for broad-match queries (a
+/// short fragment can match many entries); applied after bucket ordering so
+/// only the lowest-relevance tail is dropped, never an exact/prefix hit. The
+/// value is a generous backstop, not a wire contract beyond "bounded".
+const MAX_RESULTS: usize = 50;
```
Comment (phelma) [inline skills/visualisation/visualise/server/src/api/search.rs:25]: 🔵 **Architecture / Compatibility**

The new `MAX_RESULTS = 50` cap is applied in the shared projection chain, so it bounds title/slug/body queries too — a behavioural change (previously-unbounded → bounded) wider than the ID feature that motivates it. The placement is correct (after bucket concatenation, so only the lowest-relevance tail is dropped and exact/prefix hits are never cut), but the truncation is unsignalled: the sole consumer (`SearchResultsPanel`) renders a bare `{N} matches` from `results.length`, so a broad query that legitimately exceeds 50 displays exactly `50 matches` with no "showing first N" affordance, mis-stating the true count.

**Impact**: A corpus with >50 legitimate title matches now truncates silently, and consumers can't distinguish a capped result set from a complete one. No consumer *breaks* (there is no pagination/unbounded-count assumption), so this is a contract narrowing rather than a break — the comment's "not a wire contract beyond bounded" framing is accurate.

**Suggestion**: Acceptable as-is given the single co-shipped consumer. When a truncation indicator is wanted, the cleanest forward-compatible signal is an additive `truncated: bool` (or `total`) field on `SearchResponse` — older clients ignore unknown fields, so it stays backward-compatible.

File: skills/visualisation/visualise/server/src/api/search.rs [hunk 2]
```diff
@@ -132,7 +185,25 @@ pub(crate) async fn search(
         .into_iter()
         .flatten()
         .filter_map(|e| project(&e))
+        .take(MAX_RESULTS)
```
Comment (phelma) [inline skills/visualisation/visualise/server/src/api/search.rs:188]: 🔵 **Performance**

`.take(MAX_RESULTS)` is applied at the end of the projection chain — after the full corpus has been classified, after the matched set is bucketed, and after each bucket is sorted via `sort_by_cached_key` (which allocates a `rel_path` String key per matched entry). It therefore bounds only the response payload, JSON serialisation, and `project()` clones; it does not bound the O(corpus) classify pass, the snapshot clone, or the sort that run before truncation.

**Impact**: A broad-match query (e.g. a project-code fragment prefixing every id) still does full O(corpus) work plus a full sort over all matches before the cap drops the tail. At the current scale (one repo's local `meta/` corpus, localhost) this floor is small and acceptable.

**Suggestion**: No change needed for the expected scale — this matches the plan's documented intent. If the corpus ever grows by orders of magnitude or search becomes multi-tenant, revisit by capping the matched set per bucket before the sort (e.g. a bounded heap).

File: skills/visualisation/visualise/server/src/api/search.rs [hunk 3]
```diff
@@ -41,35 +47,82 @@ pub(crate) struct SearchResponse {
...
+/// Assumes the normalised-id shape from `WorkItemConfig::normalise_id` — bare
+/// digits or `CODE-digits` with an alphabetic prefix — so the digits are the
+/// trailing run; a future id shape with digits elsewhere would mis-key.
+fn numeric_key(id_lc: &str) -> &str {
+    id_lc
+        .rsplit(|c: char| !c.is_ascii_digit())
+        .next()
+        .unwrap_or("")
+        .trim_start_matches('0')
+}
```
Comment (phelma) [inline skills/visualisation/visualise/server/src/api/search.rs:70]: 🔵 **Architecture**

`numeric_key()` lives in `search.rs` but its correctness depends on the id-shape contract (`bare digits` or `CODE-digits` with a trailing digit run) produced by `WorkItemConfig::normalise_id` in `config.rs`. The dependency is documented and currently upheld (all `work_item_id` values route through `normalise_id`), but it is an implicit, untested invariant spanning three modules — a future id shape with digits elsewhere would silently mis-key here with no compile-time or test-time signal.

**Impact**: If `normalise_id` ever admits a new id shape, the search ranker would mis-rank with no failing test pointing back to this assumption.

**Suggestion**: Acceptable as-is for this scope given the comment. Consider a single co-located test asserting `numeric_key(normalise_id(...))` round-trips for each shape `normalise_id` can emit, so the invariant is pinned where it is relied upon rather than only described in prose.

File: skills/visualisation/visualise/server/src/api/search.rs [hunk 4]
```diff
@@ -41,35 +47,82 @@ pub(crate) struct SearchResponse {
...
+    let id_cmp: Option<&str> =
+        id_lc
+            .as_deref()
+            .map(|id| if q_is_num { numeric_key(id) } else { id });
+    let id_active = !id_q.is_empty();
+    let id_exact = id_active && id_cmp == Some(id_q);
+    let id_prefix = id_active
+        && id_cmp.is_some_and(|id| !id.is_empty() && id.starts_with(id_q));
+    let id_interior = id_active
+        && id_cmp.is_some_and(|id| !id.is_empty() && id.contains(id_q));
```
Comment (phelma) [inline skills/visualisation/visualise/server/src/api/search.rs:113]: 🔵 **Code Quality**

The `id_*` bindings carry a deliberate but unexplained asymmetry: `id_exact` relies only on `id_active`, while `id_prefix`/`id_interior` additionally re-check `!id.is_empty()`. The reason is subtle — for a numeric query `numeric_key` may reduce a real id to empty, which can never `==` a non-empty `id_q` (so `id_exact` is safe) but would spuriously `starts_with`/`contains` it (so prefix/interior need the guard). Correct, but the densely-packed booleans force the reader to reconstruct that argument.

**Impact**: A future maintainer adjusting the tiers may mis-copy the pattern (e.g. drop the `!id.is_empty()` check) without realising why it differs between branches, reintroducing a flood on empty-reduced ids.

**Suggestion**: Either add a one-line comment on why `id_exact` omits the emptiness re-check, or extract the plan's suggested `id_bucket(id_cmp, id_q) -> Option<Bucket>` helper that folds the empty-guard logic into one tested unit (mirroring the `numeric_key` extraction).

**Review by phelma — COMMENTED** [review]
## Code Review: #8 - Match work_item_id in visualiser search

**Verdict:** COMMENT

This is a tightly-scoped, high-quality change: ID matching slots into the existing four-bucket ranker with no new wire fields and no indexer change, the `numeric_key` helper is small, pure and unit-tested, and the ~500 lines of new integration tests pin *tier placement* (not mere presence) with mutation-killing rationale per test. The wire shape is unchanged so no consumer breaks. Every lens returned only minor/suggestion-grade findings — nothing blocking — and most cluster around the one genuinely cross-cutting decision: the new `MAX_RESULTS` cap, which lands in the shared projection chain and therefore quietly changes the contract for *all* query types, not just the ID feature that motivates it.

### Cross-Cutting Themes

- **The `MAX_RESULTS = 50` cap** (flagged by: architecture, compatibility, performance, test-coverage, code-quality) — Five lenses converge on the cap from different angles: it widens a previously-unbounded contract to bound title/slug/body queries too (architecture, compatibility); it truncates silently with no "showing first N" affordance, so the frontend `{N} matches` count mis-states reality past 50 (compatibility); it bounds the *response payload* but not per-query CPU — the O(corpus) classify pass and full sort still run before `.take()` (performance); the cap test does not pin that truncation happens *after* bucket concatenation (test-coverage); and the test hard-codes the literal `50` because the constant is private to the crate (code-quality, test-coverage). None is blocking, but together they argue for a follow-up: an additive `truncated`/`total` signal on the response and an order-sensitivity test.
- **The numeric ID matching logic** (flagged by: correctness, code-quality, architecture) — `numeric_key` and the `id_exact`/`id_prefix`/`id_interior` binding block carry a deliberate empty-guard asymmetry and a cross-module id-shape invariant (owned by `normalise_id` in `config.rs`). All correct, all documented in prose, but the invariant is untested where it is relied upon and the asymmetry forces re-reading.

### Strengths

- ✅ ID matching is localised to `classify()` and reuses the existing ExactSlug/Prefix/Interior tiers — no new bucket, no `IndexEntry` field, no serialised-shape change, no migration.
- ✅ The open-closed concern is handled proactively: an explicit refactor-trigger comment tells the next author to iterate candidate ID fields per tier (with a per-field numeric-reduction policy) *before* hand-copying the ladder a third time for `external_id`.
- ✅ `numeric_key` is a small, pure, borrow-only helper with a precise doc comment stating its `normalise_id`-shape assumption, covered by a focused in-module unit test including the empty/no-digit edge cases.
- ✅ Tests pin tier placement by *relative ordering* against an adjacent-bucket competitor (with deliberate mtime/rel_path control) rather than `any(...)` presence — a deleted or mis-tiered branch fails a test.
- ✅ Edge cases are well chosen and each names the branch it guards: all-zeros flood guard, `None`-id non-panic, foreign-prefixed id (`OPS-7` under an `ENG` config), project-code config, and the cap preserving the top hit.
- ✅ The lazy `body_preview` lowercasing optimisation is preserved (computed only after title/slug/id tiers miss), and the new `work_item_id` lowercasing is a no-op on the `None` majority.
- ✅ The wire shape (`docType`/`title`/`slug`/`mtimeMs`) is unchanged and still matches the frontend `SearchResult` interface exactly — backward- and forward-compatible.
- ✅ The stale `IndexEntry.work_item_id` doc comment (filename-derived) is corrected to reflect the actual frontmatter-`id:` → `normalise_id` source in the same change.

### General Findings

- 🔵 **Correctness**: For a numeric query, `numeric_key` strips *all* non-digit prefix, so `42` exact-matches `0042`, `ENG-0042` *and* `OPS-0042` simultaneously — same-numbered items across project codes collide at the exact tier. Documented and accepted; the right place to revisit is the `external_id` per-field policy already noted in the code.
- 🔵 **Correctness**: The empty-`numeric_key` guard on the prefix/interior branches is belt-and-braces (the symmetric all-zeros-*query* case is what's actually load-bearing, guarded by `id_active`). No defect; noted so the two empty cases aren't conflated in future edits.
- 🔵 **Test Coverage**: The documented-and-accepted mixed-key gap (`eng-42` does *not* match a zero-padded `ENG-0042`) has no negative test, whereas its positive siblings (`42`, `eng-0042`, `0042`) are all covered.
- 🔵 **Test Coverage**: The cap's behavioural change to the *title/slug/body* search path is asserted only indirectly (the one cap test drives truncation entirely through `work_item_id` interior matches). Low priority — the logic is field-agnostic.
- 🔵 **Code Quality**: The twelve new integration tests repeat the same ~6-line setup preamble; a small `build_work_item_state` / `seed_work_dir` helper would collapse it. Low severity given the relaxed bar for test code.
- 🔵 **Performance**: The dominant per-query cost remains `indexer.all()` cloning every `IndexEntry` (including the `frontmatter` JSON blob) under a read lock — correctly left untouched. Flag only marks where real headroom lies if search ever needs optimising.
- 🔵 **Compatibility**: Server and frontend ship as one embedded binary (`embed-dist`), so `/api/search` has no cross-version skew exposure and needs none of the `WireIndexEntry` tolerance the docs/lifecycle paths carry.

---
*Review generated by /accelerator:review-pr*

File: skills/visualisation/visualise/server/src/api/search.rs [hunk 5]
```diff
@@ -41,35 +47,82 @@ pub(crate) struct SearchResponse {
...
+    let q_is_num = q_lc.bytes().all(|b| b.is_ascii_digit());
+    let id_q: &str = if q_is_num {
+        q_lc.trim_start_matches('0')
+    } else {
+        q_lc
+    };
+    let id_cmp: Option<&str> =
+        id_lc
+            .as_deref()
+            .map(|id| if q_is_num { numeric_key(id) } else { id });
```
Comment (phelma) [inline skills/visualisation/visualise/server/src/api/search.rs:107]: 🔵 **Code Quality**

`q_is_num` and `id_q` depend only on the query, not the entry, yet they are recomputed inside `classify` for every entry in the corpus. The plan's own implementer notes suggested hoisting these query-level facts into the `search` handler and passing them in, so `classify` reads as "match this prepared query against this entry".

**Impact**: Minor — the per-entry recompute is a trivial scan of a ≤128-char string and is dwarfed by the snapshot clone, so this is a clarity choice, not a measurable cost. But colocating query-derivation with per-entry matching slightly blurs the function's responsibility.

**Suggestion**: Optionally hoist `q_is_num`/`id_q` to the call site in `search()` and pass them into `classify`. Acceptable to leave as-is given the explicit scope decision in the plan.

File: skills/visualisation/visualise/server/tests/api_search.rs [hunk 6]
```diff
@@ -428,3 +428,503 @@ async fn mixed_case_query_and_field_classify_correctly() {
...
+        results.len(),
+        50,
+        "result set must be capped at MAX_RESULTS"
+    );
+    assert_eq!(
+        results[0]["slug"].as_str(),
+        Some("exact-target"),
+        "exact hit must survive truncation at the top"
+    );
+}
```
Comment (phelma) [inline skills/visualisation/visualise/server/tests/api_search.rs:849]: 🔵 **Test Coverage**

The cap test seeds one exact-match target plus 55 interior-only entries, so the lone exact hit lands at `results[0]` whether truncation runs *before* or *after* bucket concatenation. As the plan's own verification note acknowledges, item 10 therefore does not distinguish those orderings — a mutation moving `.take(MAX_RESULTS)` ahead of `buckets.into_iter().flatten()` would still pass, yet would drop high-relevance hits in a mixed-tier flood.

**Impact**: The load-bearing claim that truncation drops only the lowest-relevance tail (never an exact/prefix hit) is not actually pinned by a test; a regression reordering the projection chain ships green.

**Suggestion**: Add the order-sensitivity test the plan describes — seed >MAX_RESULTS `Prefix` hits (slug marker `pfx-NN`) plus some `Interior` hits (`int-NN`) for one query, then assert every surviving slug carries the `Prefix` marker. This fails if truncation precedes bucket concatenation.

File: skills/visualisation/visualise/server/tests/api_search.rs [hunk 7]
```diff
@@ -428,3 +428,503 @@ async fn mixed_case_query_and_field_classify_correctly() {
...
+
+    let activity = Arc::new(Activity::new());
+    let state = AppState::build(cfg, activity).await.unwrap();
+    let v = fetch_search("/api/search?q=7", state).await;
+    let results = v["results"].as_array().unwrap();
+    assert_eq!(
+        results.len(),
+        50,
+        "result set must be capped at MAX_RESULTS"
+    );
```
Comment (phelma) [inline skills/visualisation/visualise/server/tests/api_search.rs:843]: 🔵 **Test Coverage / Code Quality**

The cap assertion hard-codes the literal `50` with a comment that it is kept in sync with `api::search::MAX_RESULTS`. The constant is private to the server crate and this integration test is a separate crate that cannot import it, so the literal is the pragmatic option — but the magic number (and the implicit `for i in 0..55` flood size, `MAX_RESULTS + 5`) silently couples to the constant with no compile-time link.

**Impact**: If `MAX_RESULTS` is raised below 55 (or the seed count drops below it), the test asserts the wrong cap or no longer overflows it, and the coupling is invisible — failures surface at runtime rather than at compile time.

**Suggestion**: Either expose `MAX_RESULTS` as `pub(crate)` plus a thin re-export so the integration test can reference it; or assert only the contractual property (`results.len() <= MAX_RESULTS` and the top hit survives); or add a one-line assertion in the in-module `mod tests` block in `search.rs` (which *can* see the private constant) pinning `MAX_RESULTS == 50`, so changing it forces a deliberate update that flags this literal.

File: skills/visualisation/visualise/server/src/indexer.rs [hunk 8]
```diff
@@ -164,9 +164,10 @@ pub struct IndexEntry {
     pub path: PathBuf,
     pub rel_path: PathBuf,
     pub slug: Option<String>,
-    /// Filename-derived work-item ID (via the configured scan regex).
-    /// `Some(id)` when the entry is a work-item and the filename matches;
-    /// `None` for non-work-item types or unmatched filenames.
+    /// Work-item identity, read from the frontmatter `id:` key and normalised
+    /// via `WorkItemConfig::normalise_id`. `Some(id)` only when the entry is a
+    /// work item whose frontmatter carries an `id:`; `None` for non-work-item
+    /// types and for work items lacking a frontmatter `id:`.
```
Comment (phelma) [inline skills/visualisation/visualise/server/src/indexer.rs:170]: 🔵 **Correctness**

The corrected doc comment states `work_item_id` is `None` only for non-work-item types and for work items "lacking a frontmatter `id:`". But the population path (`read_fm_id`) also yields `None` when an `id:` key is *present* but fails `normalise_id` shape validation (e.g. `ENG0042`, `PROJ-1.2`) — in which case it additionally logs a `tracing::warn!`.

**Impact**: The enumerated `None` conditions are incomplete; a reader relying on this comment would not expect a present-but-malformed `id:` to also produce `None`.

**Suggestion**: Extend the `None` clause, e.g. "…and for work items whose frontmatter `id:` is absent or fails `normalise_id` shape validation."
